File: str_helper.cpp
// system includes
#include <map>
#include <iomanip>                  // std::setfill

// includes
#include "str_helper.h"
#include "basic_parser/str_helper.h"

namespace basic_objects
{

namespace str_helper
{

using ::basic_parser::str_helper::write;
using ::basic_parser::str_helper::write_t;

// enums

#define TUPLE_VAL_STR(_x_)  _x_,#_x_

std::ostream & write( std::ostream & os, const weekdays_e r )
{
    typedef weekdays_e Type;
    static const std::map< Type, std::string > m =
    {
        { Type:: TUPLE_VAL_STR( MO ) },
        { Type:: TUPLE_VAL_STR( TU ) },
        { Type:: TUPLE_VAL_STR( WE ) },
        { Type:: TUPLE_VAL_STR( TH ) },
        { Type:: TUPLE_VAL_STR( FR ) },
        { Type:: TUPLE_VAL_STR( SA ) },
        { Type:: TUPLE_VAL_STR( SU ) },
    };

    auto it = m.find( r );

    static const std::string undef( "undef" );

    if( it != m.end() )
        return write( os, it->second );

    return write( os, undef );
}

std::ostream & write( std::ostream & os, const gender_e r )
{
    typedef gender_e Type;
    static const std::map< Type, std::string > m =
    {
        { Type:: TUPLE_VAL_STR( UNDEF ) },
        { Type:: TUPLE_VAL_STR( MALE ) },
        { Type:: TUPLE_VAL_STR( FEMALE ) },
        { Type:: TUPLE_VAL_STR( OTHER ) },
    };

    auto it = m.find( r );

    static const std::string undef( "undef" );

    if( it != m.end() )
        return write( os, it->second );

    return write( os, undef );
}

// objects

std::ostream & write( std::ostream & os, const TimePoint24 & l )
{
    os << std::setfill( '0' );

    os << std::setw( 2 ) << (unsigned)l.hh << ":" << std::setw( 2 ) << (unsigned)l.mm;

    return os;
}

std::ostream & write( std::ostream & os, const TimeWindow & l )
{
    os << l.from << "-" << l.to;

    return os;
}

std::ostream & write( std::ostream & os, const LocalTime & l )
{
    os << l.year << "-"
            << std::setfill( '0' ) << std::setw( 2 ) << (unsigned)l.month << "-"
            << std::setfill( '0' ) << std::setw( 2 ) << (unsigned)l.day << " "
            << std::setfill( '0' ) << std::setw( 2 ) << (unsigned)l.hh << ":"
            << std::setfill( '0' ) << std::setw( 2 ) << (unsigned)l.mm << ":"
            << std::setfill( '0' ) << std::setw( 2 ) << (unsigned)l.ss;

    return os;
}

std::ostream & write( std::ostream & os, const Weekdays & l )
{
    static const char *days[] =
    {
            "Mo", "Tu", "We", "Th", "Fr", "Sa", "So"
    };

    for( unsigned i = 0; i < 7; ++i )
    {
        auto e = 1 << i;

        if( l.mask & e )
        {
            os << days[i];
        }
        else
        {
            os << "__";
        }
    }

    return os;
}

std::string to_string_YYYYMMDD( const Date & l )
{
    std::ostringstream os;

    os << std::setfill( '0' );

    os << std::setw( 4 ) << (unsigned) l.year << std::setw( 2 ) << (unsigned) l.month << std::setw( 2 ) << (unsigned) l.day;

    return os.str();
}

std::string to_string_HHMM( const TimePoint24 & l )
{
    std::ostringstream os;

    os << std::setfill( '0' ) << std::setw( 2 ) << (unsigned) l.hh << std::setfill( '0' ) << std::setw( 2 ) << (unsigned) l.mm;

    return os.str();
}

std::ostream & write( std::ostream & os, const TimeRange & l )
{
    os << l.from;
    os << "-";
    os << l.to;

    return os;
}

std::ostream & write( std::ostream & os, const LocalTimeRange & l )
{
    write( os, l.from );
    os << "-";
    write( os, l.to );

    return os;
}

std::ostream & write( std::ostream & os, const Date & l )
{
    os << std::setfill( '0' );

    os << std::setw( 4 ) << (unsigned) l.year << std::setw( 2 ) << (unsigned) l.month << std::setw( 2 ) << (unsigned) l.day;

    return os;
}

std::ostream & write( std::ostream & os, const Email & l )
{
    os << l.email;

    return os;
}

// base messages

// messages

} // namespace str_helper

} // namespace basic_objects

File: str_helper.php
<?php

namespace basic_objects;


// includes
require_once __DIR__.'/../basic_parser/str_helper.php';

// enums

function to_string__weekdays_e( $r )
{
    $map = array
    (
        weekdays_e__MO => 'MO',
        weekdays_e__TU => 'TU',
        weekdays_e__WE => 'WE',
        weekdays_e__TH => 'TH',
        weekdays_e__FR => 'FR',
        weekdays_e__SA => 'SA',
        weekdays_e__SU => 'SU',
    );

    if( array_key_exists( $r, $map ) )
    {
        return $map[ $r ];
    }

    return '?';
}

function to_string__gender_e( $r )
{
    $map = array
    (
        gender_e__UNDEF => 'UNDEF',
        gender_e__MALE => 'MALE',
        gender_e__FEMALE => 'FEMALE',
        gender_e__OTHER => 'OTHER',
    );

    if( array_key_exists( $r, $map ) )
    {
        return $map[ $r ];
    }

    return '?';
}

// objects

function to_string__TimePoint24( & $obj )
{
    return sprintf( "%02d:%02d", $obj->hh, $obj->mm );
}

function to_string__TimeWindow( & $obj )
{
    return to_string__TimePoint24( $obj->from ) . "-" . to_string__TimePoint24( $obj->to );
}

function to_string__LocalTime( & $obj )
{
    return sprintf( "%04d-%02d-%02d %02d:%02d:%02d", $obj->year, $obj->month, $obj->day, $obj->hh, $obj->mm, $obj->ss );
}

function to_string__Weekdays( & $r )
{
    $res = "(";

    $res .= " mask=" . \basic_parser\to_string__int( $r->mask );

    $res .= ")";

    return $res;
}

function to_string__TimeRange( $obj )
{
    return "from=" . $obj->from . " to=" . $obj->to;
}

function to_string__LocalTimeRange( $obj )
{
    return to_string__LocalTime( $obj->from ) . " - " . to_string__LocalTime( $obj->to );
}

function to_string__Date( & $obj )
{
    return sprintf( "%04d-%02d-%02d", $obj->year, $obj->month, $obj->day );
}

function to_string__Email( & $obj )
{
    return $obj->email;
}

// base messages

// messages

// generic

function to_string( $obj )
{
    $handler_map = array(
        // objects
        'basic_objects\TimePoint24'         => 'to_string__TimePoint24',
        'basic_objects\TimeWindow'         => 'to_string__TimeWindow',
        'basic_objects\LocalTime'         => 'to_string__LocalTime',
        'basic_objects\Weekdays'         => 'to_string__Weekdays',
        'basic_objects\TimeRange'         => 'to_string__TimeRange',
        'basic_objects\LocalTimeRange'         => 'to_string__LocalTimeRange',
        'basic_objects\Date'         => 'to_string__Date',
        'basic_objects\Email'         => 'to_string__Email',
        // messages
    );

    $type = get_class( $obj );

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\basic_objects\\' . $handler_map[ $type ];
        return $func( $obj );
    }

    return NULL;
}

# namespace_end basic_objects


?>
